src/new-storage/Storage.js

```diff
@@ -173,55 +152,191 @@ class Storage extends EventEmitter {
+                console.warn(e)
+                resultStream.push(null)
+            })
 
         return resultStream
     }
 
-    _fetchBetweenMessageRefsForPublisher(streamId, partition, fromTimestamp, fromSequenceNo, toTimestamp, toSequenceNo, publisherId, msgChainId) {
+    _fetchBetweenTimestamps(streamId, partition, fromTimestamp, toTimestamp) {
         const resultStream = this._createResultStream()
 
-        // Cassandra doesn't allow ORs in WHERE clause so we need to do 3 queries.
-        // Once a range (id/partition/ts/sequence_no) has been selected in Cassandra, filtering it by publisher_id requires to ALLOW FILTERING.
-        const query1 = 'SELECT payload FROM stream_data WHERE id = ? AND partition = ? AND ts = ? AND sequence_no >= ? AND publisher_id = ? '
-            + 'AND msg_chain_id = ? ORDER BY ts ASC, sequence_no ASC ALLOW FILTERING'
-        const query2 = 'SELECT payload FROM stream_data WHERE id = ? AND partition = ? AND ts > ? AND ts < ? AND publisher_id = ? '
-            + 'AND msg_chain_id = ? ORDER BY ts ASC, sequence_no ASC ALLOW FILTERING'
-        const query3 = 'SELECT payload FROM stream_data WHERE id = ? AND partition = ? AND ts = ? AND sequence_no <= ? AND publisher_id = ? '
-            + 'AND msg_chain_id = ? ORDER BY ts ASC, sequence_no ASC ALLOW FILTERING'
-        const queryParams1 = [streamId, partition, fromTimestamp, fromSequenceNo, publisherId, msgChainId]
-        const queryParams2 = [streamId, partition, fromTimestamp, toTimestamp, publisherId, msgChainId]
-        const queryParams3 = [streamId, partition, toTimestamp, toSequenceNo, publisherId, msgChainId]
-        const stream1 = this._queryWithStreamingResults(query1, queryParams1)
-        const stream2 = this._queryWithStreamingResults(query2, queryParams2)
-        const stream3 = this._queryWithStreamingResults(query3, queryParams3)
-
-        pump(
-            merge2(stream1, stream2, stream3),
-            resultStream,
-            (err) => {
-                if (err) {
-                    console.error('pump finished with error', err)
-                    resultStream.push(null)
+        const query = 'SELECT payload FROM stream_data_new WHERE '
+            + 'stream_id = ? AND partition = ? AND bucket_id IN ? AND ts >= ? AND ts <= ?'
+
+        Promise.all([
+            this.bucketManager.getLastBuckets(streamId, partition, 1, fromTimestamp),
+            this.bucketManager.getLastBuckets(streamId, partition, 1, toTimestamp),
+        ]).then((results) => {
+            const date1 = results[0][0].dateCreate
+            const date2 = results[1][0].dateCreate
+            return [Math.min(date1, date2), Math.max(date1, date2)]
+        }).then(([startBucketDate, endBucketDate]) => {
+            return this.bucketManager.getBucketsByTimestamp(streamId, partition, startBucketDate, endBucketDate)
+        }).then((buckets) => {
+            const bucketsForQuery = []
+
+            for (let i = 0; i < buckets.length; i++) {
```

Comment:
This code segment is repeated a lot. Could maybe wrap as a helper function? E.g. `resultSetToBucketIds(buckets)` or something.